File: logic_implementation/lc_273_Integer_to_English_Words/python/solution.py
import math
class Solution:

    def __init__(self):
        self.under20 = {
            1: 'One',
            2: 'Two',
            3: 'Three',
            4: 'Four',
            5: 'Five',
            6: 'Six',
            7: 'Seven',
            8: 'Eight',
            9: 'Nine',
            10: 'Ten',
            11: 'Eleven',
            12: 'Twelve',
            13: 'Thirteen',
            14: 'Fourteen',
            15: 'Fifteen',
            16: 'Sixteen',
            17: 'Seventeen',
            18: 'Eighteen',
            19: 'Nineteen',
        }
        self.under100 = {
            2: 'Twenty',
            3: 'Thirty',
            4: 'Forty',
            5: 'Fifty',
            6: 'Sixty',
            7: 'Seventy',
            8: 'Eighty',
            9: 'Ninety',
        }

        self.multiplerName = {
            1: ' Hundred',
            2: ' Thousand',
            3: ' Million',
            4: ' Billion',
        }
        self.multiplerUnit = {
            1: 100,
            2: 1000,
            3: 1000000,
            4: 1000000000,
        }
    def numberToWords(self, num: int) -> str:
        return self.toWords('', num)

    def toWords(self, prev: str, num: int) -> str:
        if num == 0:
            return 'Zero' if not prev else prev
        if num < 20:
            return self.under20[num] if not prev else prev + ' ' + self.under20[num]
        if num < 100:
            answer = self.toWords(self.under100[num//10], num % 10)
            return answer if not prev else prev + ' ' + answer

        index = self.getMultiplerIndex(num)
        answer = self.toWords(prev, num//self.multiplerUnit[index]) + self.multiplerUnit[index]
        return self.toWords(answer, num % self.multiplerUnit[index])

    def getMultiplerIndex(self, num: int) -> int:
        n = len(str(num))
        return math.ceil(n/3)

File: logic_implementation/lc_273_Integer_to_English_Words/java/Solution.java
public class Solution{
    static String[] under20 = {
            "Zero", "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten",
            "Eleven", "Twelve", "Thirteen", "Fourteen", "Fifteen", "Sixteen", "Seventeen", "Eighteen", "Nineteen"};
    static String[] under100 = {
            "", "Ten", "Twenty", "Thirty", "Fourty", "Fifty", "Sixty", "Seventy", "Eighty", "Ninety"
    };
    static String[] multiplierName = {"", " Hundred", " Thousand", " Million", " Billion"};
    static int[] multiplierUnit    = {1,   100,         1000,       1000000,    1000000000};
    public String numberToWords(int num) {
        return toWords("", num);
    }

    String toWords(String prev, int num) {
        if (num == 0) {
            return prev.length() == 0 ? "Zero" : prev;
        }
        if (num < 20) {
            return prev.length() == 0 ? under20[num] : prev + " " + under20[num];
        }
        if (num < 100) {
            String answer = toWords(under100[num/10], num % 10);
            return prev.length() == 0 ? answer : prev + " " + answer;
        }
        if (num < 1000) {
            String answer = toWords(prev, num / 100) + " Hundred";
            return toWords(answer, num % 100);
        }
        int n = Integer.toString(num).length();
        int index = (int)Math.ceil(n/3.0);

        String answer = toWords(prev, num / multiplierUnit[index]) + multiplierName[index];
        return toWords(answer, num % multiplierUnit[index]);
    }

    public static void main(String[] args) {
        t(12345);
        t(123456780);
    }
    static void t(int n) {
        System.out.printf("%d -> %s\n", n, new Solution().numberToWords(n));
    }
}